### ISSUE TITLE
Thanks - and a note for users who may be struggling to get this working

### ISSUE THREAD

**Issue by Majestic7979** [l4rm4nd/F2BFilters#4]
Hi! I found instructions on your blog, I was able to get this working but I had to use INPUT chain instead of DOCKER-USER. On DOCKER-USER chain the rule was added to iptables, but it was not enforcing the ban even though fail2ban logs on docker was showing as banned, I changed the DOCKER-USER to INPUT and now it's working as expected. Thanks a lot for sharing your knowledge to help make things more secure.

**Comment by l4rm4nd** [comment]
Hi @Majestic7979 ,

glad you got it up and running.

**tldr**: The INPUT chain usually does not work for Docker as it is bypassed. You must use the DOCKER-USER chain. I highly recommend validating your setup and that IPs are effectively banned. Also, my blog and this GH repo requires iptables due to the utilized string matching capability. Nftables, which comes in modern distros, does not support this.

---------

Just a side note in general. Docker bypasses the INPUT rules spawned by iptables or nftables. So even if you create restrictive firewall rules, spawned up containers that map ports onto your host are often accessible still.

So usually you must use the DOCKER-USER chain, as the firewalls rules are otherwise not effecitve. If this is not the case right now, it's likely due to your infrastructure/network setup. I highly recommend validating whether IP bans indeed work.

Also, when working with fail2ban and local iptables bans, it is crucial to use iptables and not nftables. The blog post as well as this repository here is based on using iptables' string matching capabilities. This feature is only available on iptables and not nftables. However, many newer distros come with nftables only. So the ip bans will not work, especially if you are using a setup of dockerized reverse proxies. 

The reason for this lies in the fact that iptables can only see the IP packet's source IP address. If you are using a reverse proxy, the source IP of all your packets is actually the IP of the reverse proxy. So fail2ban may go ahead and ban your own infrastructure host, the reverse proxy. Iptables is not able to inspect the IP packet itself and parse the relevant headers such as X-Forwarded-For or X-Real-IP. These are the ones, typically, which are propagated with the IP address of the real website visitor of your services that tries to cause harm.

**Comment by Majestic7979** [comment]
Hey, thanks for explaining. Please hear me out. Trust me, it's working with INPUT chain...

In the DOCKER-USER chain the fail2ban logs show it says it banned, I can see the rule in iptables (I'm using legacy iptables, not nftables) but when I try to access using my 5G connection it still allows full access. As soon as I moved to INPUT the access completed stopped.

The page never loads. It starts loading again after the 10 minutes. So I don't know what differs in my setup compared to yours but what you said above does not apply to me, and potentially does not apply to other users. I am running the latest version of docker on the Raspberry Pi OS.

I use the Raspberry Pi 4. That's where my traefik sits (running from docker) alongside fail2ban also running from docker. That's the entrypoint to my network. All other services run on a separate server which has a different IP than the Raspberry Pi running traefik and fail2ban.

If you don't believe me, I'll be happy to arrange a screenshare session with you to prove it. Using DOCKER-USER adds the rules but doesn't do anything. User INPUT does in fact block. I can reproduce this effect over and over and over...

**Comment by l4rm4nd** [comment]
> Hey, thanks for explaining. Please hear me out. Trust me, it's working with INPUT chain...
> 
> In the DOCKER-USER chain the fail2ban logs show it says it banned, I can see the rule in iptables (I'm using legacy iptables, not nftables) but when I try to access using my 5G connection it still allows full access. As soon as I moved to INPUT the access completed stopped.
> 
> The page never loads. It starts loading again after the 10 minutes. So I don't know what differs in my setup compared to yours but what you said above does not apply to me, and potentially does not apply to other users. I am running the latest version of docker on the Raspberry Pi OS.
> 
> I use the Raspberry Pi 4. That's where my traefik sits (running from docker) alongside fail2ban also running from docker. That's the entrypoint to my network. All other services run on a separate server which has a different IP than the Raspberry Pi running traefik and fail2ban.
> 
> If you don't believe me, I'll be happy to arrange a screenshare session with you to prove it. Using DOCKER-USER adds the rules but doesn't do anything. User INPUT does in fact block. I can reproduce this effect over and over and over...

All good m8, I believe you.

Just wanted to state that this is not the default behaviour, as Docker in its default installation will bypass iptables rules.

Do you run Traefik in host network mode or bridge?

**Comment by Majestic7979** [comment]
Thanks for believing what I said.

I run traefik in host mode. I use bind mounts for data persistence.

**Comment by l4rm4nd** [comment]
> I run traefik in host mode.

That's likely the reason why your INPUT chain works. You do not rely on Docker networks like bridge.

**Comment by Majestic7979** [comment]
> > I run traefik in host mode.
> 
> That's likely the reason why your INPUT chain works. You do not rely on Docker networks like bridge.

Right, I only expose ports for some containers where I need to tweak the external port number. For traefik this is not really required, I want it using the standard 80 and 443. Hopefully this info is at least useful to anyone coming here while troubleshooting.

**Comment by l4rm4nd** [comment]
> > > I run traefik in host mode.
> > 
> > That's likely the reason why your INPUT chain works. You do not rely on Docker networks like bridge.
> 
> Right, I only expose ports for some containers where I need to tweak the external port number. For traefik this is not really required, I want it using the standard 80 and 443. Hopefully this info is at least useful to anyone coming here while troubleshooting.

FYI, you can run Traefik in Docker bridge network and manually map the TCP ports 80 and 443 on the host. No need to use the host network.

But it's your infra and it seems to work now with fail2ban. So let's leave it there. 

Your post sure will help others. Thanks!
